File: agent/handlers/dockerHandler.js
const { spawn } = require('child_process');
const os = require('os');
let pty;
try {
    pty = require('node-pty');
} catch (e) {
    console.warn('node-pty not found, falling back to basic child_process. Interactive terminal features (colors, resizing) will be limited.');
}

const logStreams = new Map(); // containerId -> process
const terminalSessions = new Map(); // containerId -> process (pty or spawn)

/**
 * Detect available shell in container
 * @param {string} containerId - Docker container ID
 * @returns {Promise<string>} - Path to available shell
 */
async function detectShell(containerId) {
    const { exec } = require('child_process');
    const util = require('util');
    const execPromise = util.promisify(exec);

    // Try shells in order of preference
    const shells = ['sh', 'bash', 'ash'];

    for (const shell of shells) {
        try {
            // Try to execute the shell with --version or -c 'echo test'
            const { stdout } = await execPromise(
                `docker exec ${containerId} ${shell} -c "echo ok"`,
                { timeout: 2000 }
            );
            if (stdout.trim() === 'ok') {
                console.log(`Detected shell for ${containerId}: ${shell}`);
                return shell;
            }
        } catch (e) {
            // Shell not found or failed, try next
            continue;
        }
    }

    // Fallback to sh (most universal)
    console.warn(`No shell detected for ${containerId}, using sh as fallback`);
    return 'sh';
}

/**
 * Start streaming logs for a container
 * @param {Object} socket - Socket.io socket
 * @param {string} containerId - Docker container ID
 */
function startLogs(socket, containerId) {
    if (logStreams.has(containerId)) {
        stopLogs(containerId);
    }

    console.log(`Starting logs for ${containerId}`);

    const logs = spawn('docker', ['logs', '-f', '--tail', '100', containerId]);
    logStreams.set(containerId, logs);

    logs.stdout.on('data', (data) => {
        socket.emit('docker:logs:data', { containerId, data: data.toString(), type: 'stdout' });
    });

    logs.stderr.on('data', (data) => {
        socket.emit('docker:logs:data', { containerId, data: data.toString(), type: 'stderr' });
    });

    logs.on('close', (code) => {
        console.log(`Logs stream for ${containerId} exited with code ${code}`);
        logStreams.delete(containerId);
        socket.emit('docker:logs:end', { containerId, code });
    });

    logs.on('error', (err) => {
        console.error(`Logs stream error for ${containerId}:`, err);
        socket.emit('docker:logs:error', { containerId, error: err.message });
    });
}

/**
 * Stop streaming logs for a container
 * @param {string} containerId 
 */
function stopLogs(containerId) {
    const process = logStreams.get(containerId);
    if (process) {
        process.kill();
        logStreams.delete(containerId);
        console.log(`Stopped logs for ${containerId}`);
    }
}

/**
 * Start interactive terminal for a container
 * @param {Object} socket - Socket.io socket
 * @param {string} containerId - Docker container ID
 */
async function startTerminal(socket, containerId) {
    if (terminalSessions.has(containerId)) {
        stopTerminal(containerId);
    }

    console.log(`Starting terminal for ${containerId}`);

    try {
        // Detect available shell
        const shell = await detectShell(containerId);

        if (pty) {
            // Use node-pty for full terminal experience
            try {
                const ptyProcess = pty.spawn('docker', ['exec', '-it', containerId, shell], {
                    name: 'xterm-color',
                    cols: 80,
                    rows: 30,
                    cwd: process.env.HOME,
                    env: process.env
                });

                terminalSessions.set(containerId, { process: ptyProcess, type: 'pty' });

                ptyProcess.onData((data) => {
                    socket.emit('docker:terminal:data', { containerId, data });
                });

                ptyProcess.onExit(({ exitCode }) => {
                    console.log(`Terminal (PTY) for ${containerId} exited with code ${exitCode}`);
                    terminalSessions.delete(containerId);
                    socket.emit('docker:terminal:exit', { containerId, exitCode });
                });

                console.log(`Started PTY terminal for ${containerId} with ${shell}`);

            } catch (error) {
                console.error(`Failed to start PTY terminal for ${containerId}:`, error);
                socket.emit('docker:terminal:error', {
                    containerId,
                    error: `Failed to start terminal: ${error.message}`
                });
            }
        } else {
            // Fallback to child_process (without node-pty)
            try {
                // Use -i only (no -t) to avoid TTY issues without node-pty
                const cmd = spawn('docker', ['exec', '-i', containerId, shell]);

                terminalSessions.set(containerId, { process: cmd, type: 'spawn' });

                cmd.stdout.on('data', (data) => {
                    socket.emit('docker:terminal:data', { containerId, data: data.toString() });
                });

                cmd.stderr.on('data', (data) => {
                    socket.emit('docker:terminal:data', { containerId, data: data.toString() });
                });

                cmd.on('close', (code) => {
                    console.log(`Terminal (spawn) for ${containerId} exited with code ${code}`);
                    terminalSessions.delete(containerId);
                    socket.emit('docker:terminal:exit', { containerId, exitCode: code });
                });

                cmd.on('error', (err) => {
                    console.error(`Failed to start spawn terminal for ${containerId}:`, err);
                    socket.emit('docker:terminal:error', {
                        containerId,
                        error: `Terminal error: ${err.message}`
                    });
                });

                console.log(`Started spawn terminal for ${containerId} with ${shell} (basic mode)`);

            } catch (error) {
                console.error(`Failed to start fallback terminal for ${containerId}:`, error);
                socket.emit('docker:terminal:error', {
                    containerId,
                    error: `Failed to start terminal: ${error.message}`
                });
            }
        }
    } catch (error) {
        console.error(`Failed to detect shell for ${containerId}:`, error);
        socket.emit('docker:terminal:error', {
            containerId,
            error: `Container may not have a shell available: ${error.message}`
        });
    }
}

/**
 * Write data to terminal session
 * @param {string} containerId 
 * @param {string} data 
 */
function writeTerminal(containerId, data) {
    const session = terminalSessions.get(containerId);
    if (session) {
        if (session.type === 'pty') {
            session.process.write(data);
        } else {
            session.process.stdin.write(data);
        }
    }
}

/**
 * Resize terminal session
 * @param {string} containerId 
 * @param {number} cols 
 * @param {number} rows 
 */
function resizeTerminal(containerId, cols, rows) {
    const session = terminalSessions.get(containerId);
    if (session && session.type === 'pty') {
        try {
            session.process.resize(cols, rows);
        } catch (e) {
            console.error(`Error resizing terminal for ${containerId}:`, e);
        }
    }
}

/**
 * Stop terminal session
 * @param {string} containerId 
 */
function stopTerminal(containerId) {
    const session = terminalSessions.get(containerId);
    if (session) {
        if (session.type === 'pty') {
            session.process.kill();
        } else {
            session.process.kill();
        }
        terminalSessions.delete(containerId);
        console.log(`Stopped terminal for ${containerId}`);
    }
}

module.exports = {
    startLogs,
    stopLogs,
    startTerminal,
    writeTerminal,
    resizeTerminal,
    stopTerminal,
    deployCompose
};

/**
 * Deploy Docker Compose stack
 * @param {string} composeContent 
 * @returns {Promise<{message: string, output: string}>}
 */
async function deployCompose(composeContent) {
    const fs = require('fs').promises;
    const path = require('path');
    const { exec } = require('child_process');

    const tempDir = os.tmpdir();
    const filePath = path.join(tempDir, `docker-compose-${Date.now()}.yml`);

    try {
        await fs.writeFile(filePath, composeContent);

        return new Promise((resolve, reject) => {
            // Try 'docker compose' (v2)
            const command = `docker compose -f "${filePath}" up -d --remove-orphans`;

            console.log(`Executing: ${command}`);

            exec(command, (error, stdout, stderr) => {
                // Clean up file (async, don't wait)
                fs.unlink(filePath).catch(e => console.error('Failed to cleanup temp file:', e));

                if (error) {
                    console.warn('docker compose v2 failed, trying v1...', error.message);

                    // Fallback to 'docker-compose' (v1)
                    const fallbackCommand = `docker-compose -f "${filePath}" up -d --remove-orphans`;

                    exec(fallbackCommand, (err2, stdout2, stderr2) => {
                        if (err2) {
                            reject({ message: err2.message, output: stderr2 || err2.message });
                        } else {
                            resolve({ message: 'Deployed successfully (v1)', output: stdout2 });
                        }
                    });
                    return;
                }

                resolve({ message: 'Deployed successfully', output: stdout });
            });
        });
    } catch (err) {
        throw new Error(`Failed to process compose file: ${err.message}`);
    }
}

File: agent/src/handlers/docker_handler.cpp
#include "nexus/handlers/docker_handler.h"
#include "nexus/utils/logger.h"
#include <cstdio>
#include <cstdlib>
#include <array>
#include <memory>
#include <sstream>
#include <unistd.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <thread>

namespace nexus {
namespace handlers {

DockerHandler::DockerHandler() {
}

std::string DockerHandler::executeCommand(const std::string& command) {
    std::array<char, 128> buffer;
    std::string result;
    
    FILE* pipe = popen(command.c_str(), "r");
    if (!pipe) {
        Logger::getInstance().error("Failed to execute command: {}", command);
        return "";
    }
    
    while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
        result += buffer.data();
    }
    
    int status = pclose(pipe);
    
    if (status != 0) {
        Logger::getInstance().warn("Command exited with status {}: {}", status, command);
    }
    
    return result;
}

DockerControlResult DockerHandler::startContainer(const std::string& containerId) {
    Logger::getInstance().info("Starting container: {}", containerId);
    
    std::string command = "docker start " + containerId + " 2>&1";
    std::string output = executeCommand(command);
    
    DockerControlResult result;
    result.success = !output.empty() ? output.find(containerId) != std::string::npos : false;
    // docker start simply prints the containerId on success
    // On error, it prints "Error response from daemon..."
    if (output.find("Error") != std::string::npos || output.find("No such container") != std::string::npos) {
        result.success = false;
    } else {
        result.success = true;
    }
    
    result.message = result.success ? "Container started successfully" : ("Failed to start container: " + output);
    result.output = output;
    
    return result;
}

DockerControlResult DockerHandler::stopContainer(const std::string& containerId) {
    Logger::getInstance().info("Stopping container: {}", containerId);
    
    std::string command = "docker stop " + containerId + " 2>&1";
    std::string output = executeCommand(command);
    
    DockerControlResult result;
    if (output.find("Error") != std::string::npos || output.find("No such container") != std::string::npos) {
        result.success = false;
    } else {
        result.success = true;
    }
    
    result.message = result.success ? "Container stopped successfully" : ("Failed to stop container: " + output);
    result.output = output;
    
    return result;
}

DockerControlResult DockerHandler::restartContainer(const std::string& containerId) {
    Logger::getInstance().info("Restarting container: {}", containerId);
    
    std::string command = "docker restart " + containerId + " 2>&1";
    std::string output = executeCommand(command);
    
    DockerControlResult result;
    if (output.find("Error") != std::string::npos || output.find("No such container") != std::string::npos) {
        result.success = false;
    } else {
        result.success = true;
    }
    
    result.message = result.success ? "Container restarted successfully" : ("Failed to restart container: " + output);
    result.output = output;
    
    return result;
}

DockerControlResult DockerHandler::removeContainer(const std::string& containerId) {
    Logger::getInstance().info("Removing container: {}", containerId);
    
    std::string command = "docker rm -f " + containerId + " 2>&1";
    std::string output = executeCommand(command);
    
    DockerControlResult result;
    if (output.find("Error") != std::string::npos || output.find("No such container") != std::string::npos) {
        result.success = false;
    } else {
        result.success = true;
    }
    
    result.message = result.success ? "Container removed successfully" : ("Failed to remove container: " + output);
    result.output = output;
    
    return result;
}

DockerControlResult DockerHandler::removeNetwork(const std::string& networkId) {
    Logger::getInstance().info("Removing network: {}", networkId);
    
    std::string command = "docker network rm " + networkId + " 2>&1";
    std::string output = executeCommand(command);
    
    DockerControlResult result;
    result.success = !output.empty() && output.find("Error") == std::string::npos;
    result.message = result.success ? "Network removed successfully" : ("Failed to remove network: " + output);
    result.output = output;
    
    return result;
}
DockerControlResult DockerHandler::createContainer(
    const std::string& image,
    const std::string& name,
    const std::string& ports,
    const std::string& env,
    const std::string& restart,
    const std::string& command) {
    
    Logger::getInstance().info("Creating container from image: {}", image);
    
    std::ostringstream cmd;
    cmd << "docker run -d";
    
    if (!name.empty()) {
        cmd << " --name " << name;
    }
    
    if (!restart.empty() && restart != "no") {
        cmd << " --restart " << restart;
    }
    
    // Parse ports (comma-separated)
    if (!ports.empty()) {
        std::istringstream portsStream(ports);
        std::string port;
        while (std::getline(portsStream, port, ',')) {
            if (!port.empty()) {
                cmd << " -p " << port;
            }
        }
    }
    
    // Parse environment variables (comma-separated)
    if (!env.empty()) {
        std::istringstream envStream(env);
        std::string envVar;
        while (std::getline(envStream, envVar, ',')) {
            if (!envVar.empty()) {
                cmd << " -e " << envVar;
            }
        }
    }
    
    cmd << " " << image;
    
    if (!command.empty()) {
        cmd << " " << command;
    }
    
    std::string output = executeCommand(cmd.str() + " 2>&1");
    
    DockerControlResult result;
    // docker run -d prints the partial container ID on success
    // On error, it prints "docker: Error response from daemon..." or similar
    if (output.find("Error") != std::string::npos || output.find("docker:") != std::string::npos) {
        result.success = false;
    } else {
        result.success = true;
    }
    
    result.message = result.success ? "Container created successfully" : ("Failed to create container: " + output);
    result.output = output;
    
    return result;
}

bool DockerHandler::startLogs(const std::string& containerId,
                              std::function<void(const std::string&, const std::string&)> callback) {
    Logger::getInstance().info("Starting log stream for container: {}", containerId);
    
    stopLogs(containerId);
    
    int pipefd[2];
    if (pipe(pipefd) == -1) {
        Logger::getInstance().error("Failed to create pipe");
        return false;
    }
    
    pid_t pid = fork();
    if (pid == -1) {
        Logger::getInstance().error("Failed to fork");
        close(pipefd[0]);
        close(pipefd[1]);
        return false;
    }
    
    if (pid == 0) { // Child
        close(pipefd[0]); // Close read end
        dup2(pipefd[1], STDOUT_FILENO);
        dup2(pipefd[1], STDERR_FILENO);
        close(pipefd[1]);
        
        execlp("docker", "docker", "logs", "-f", "--tail", "100", containerId.c_str(), nullptr);
        exit(1);
    }
    
    // Parent
    close(pipefd[1]); // Close write end
    
    // Store stream info
    LogStream stream;
    stream.pid = pid;
    stream.callback = callback;
    log_streams_[containerId] = stream;
    
    // Start reader thread
    std::thread([this, containerId, fd = pipefd[0], cb = callback]() {
        std::array<char, 1024> buffer;
        while (true) {
            ssize_t bytes = read(fd, buffer.data(), buffer.size() - 1);
            if (bytes <= 0) break;
            
            buffer[bytes] = '\0';
            cb(containerId, std::string(buffer.data()));
        }
        close(fd);
    }).detach();
    
    return true;
}

void DockerHandler::stopLogs(const std::string& containerId) {
    auto it = log_streams_.find(containerId);
    if (it != log_streams_.end()) {
        kill(it->second.pid, SIGTERM);
        waitpid(it->second.pid, nullptr, 0); // Avoid zombies
        log_streams_.erase(it);
        Logger::getInstance().info("Stopped log stream for container: {}", containerId);
    }
}

bool DockerHandler::startTerminal(const std::string& containerId,
                                  std::function<void(const std::string&)> callback) {
    Logger::getInstance().info("Starting terminal for container: {}", containerId);
    
    // Stop existing terminal if any
    stopTerminal(containerId);
    
    // TODO: Implement terminal using fork/exec with PTY
    // For now, just log that it's not implemented
    Logger::getInstance().warn("Terminal not yet implemented in C++");
    
    return false;
}

void DockerHandler::writeTerminal(const std::string& containerId, const std::string& data) {
    auto it = terminal_sessions_.find(containerId);
    if (it != terminal_sessions_.end()) {
        // TODO: Write to terminal stdin
        write(it->second.stdin_fd, data.c_str(), data.length());
    }
}

void DockerHandler::stopTerminal(const std::string& containerId) {
    auto it = terminal_sessions_.find(containerId);
    if (it != terminal_sessions_.end()) {
        // TODO: Kill the terminal process
        terminal_sessions_.erase(it);
        Logger::getInstance().info("Stopped terminal for container: {}", containerId);
    }
}

DockerControlResult DockerHandler::deployCompose(const std::string& composeContent) {
    Logger::getInstance().info("Deploying Docker Compose stack");
    
    // Write compose content to temp file
    std::string tempFile = "/tmp/docker-compose-" + std::to_string(time(nullptr)) + ".yml";
    
    FILE* file = fopen(tempFile.c_str(), "w");
    if (!file) {
        DockerControlResult result;
        result.success = false;
        result.message = "Failed to create temporary compose file";
        return result;
    }
    
    fprintf(file, "%s", composeContent.c_str());
    fclose(file);
    
    // Try docker compose v2
    std::string command = "docker compose -f " + tempFile + " up -d --remove-orphans 2>&1";
    std::string output = executeCommand(command);
    
    // Clean up temp file
    unlink(tempFile.c_str());
    
    DockerControlResult result;
    result.success = output.find("error") == std::string::npos;
    result.message = result.success ? "Deployed successfully" : "Deployment failed";
    result.output = output;
    
    return result;
}

} // namespace handlers
} // namespace nexus
